src/Resources/contao/library/Contao/User.php

```diff
@@ -95,7 +98,7 @@
  *
  * @author Leo Feyer <https://github.com/leofeyer>
  */
-abstract class User extends \System
+abstract class User extends System implements AdvancedUserInterface, EncoderAwareInterface, \Serializable
```

Comment:
This causes problems. What Symfony Security means with their `UserInterface` is not the user entity, but the user identity. Making your user entity also the identity causes problems, when stored roles change (and other stuff). I would like to explain this to you via mumble.

Reply:
It would be better to publicly explain, because then everyone can follow and find about it

Reply:
What I actually read on the documentation doesn't sound so wrong:
> Regardless of how your user are loaded or where they come from (a database, configuration, web service, etc), you will have a class that implements this interface. Objects that implement this interface are created and loaded by different objects that implement UserProviderInterface

Reply:
The symfony docs do not go very deep on this and dont address issues like role change. You can google for these kind of issues and will find solutions like removing the login token after you updated the user roles, but all this can be avoided by using a separate object "Identity" that is derived from your actual user entity.

> It would be better to publicly explain, because then everyone can follow and find about it

I never said, i wont do it publicly, but just that writing paragraphs here isnt the correct medium, to explain a design flaw

Reply:
I've never experienced issues like that and it's for sure not my first project with Symfony security.  Even the everywhere used [`FosUserBundle`](https://github.com/FriendsOfSymfony/FOSUserBundle/blob/master/Model/User.php) also makes no distinction between entity and identity.
Correct me, if I'm wrong.

Reply:
https://github.com/symfony/symfony/issues/12025
https://github.com/symfony/symfony/pull/24331

These solutions dont handle all edge cases (and arent even merged yet)

Reply:
Thanks! I'll have a deeper look into the issues 👍

Reply:
I can't think of any way where the Symfony user is not our legacy Contao user class, or we will run into a ton of other problems. I quickly read up on the issues but couldn't find any relation to entites or models?

Reply:
> I can't think of any way where the Symfony user is not our legacy Contao user class, or we will run into a ton of other problems.

That is not the point of separating out the identity from the user entity. The identity is an adapter of the user entity to integrate with the symfony security. So it is strictly bound to contaos user entity. It also keeps your user entity clean (since you dont have possible empty implementations of unused interface methods there). But the main point is that you can tailor the identitys runtime lifecycle to play nice with symfony without hacking your user entities lifecycle (especially when you want to use doctrine for the user entity).

> I quickly read up on the issues but couldn't find any relation to entites or models?

The problem comes from serialization. You dont know if your user entity comes from the DB or the session. If you use an adapter/proxy identity you know that the Identity will be always "complete" (because it can be serialized and unserialized) and always comes from the session or runtime and never from the DB. Think of the identity as a value object referencing your user entity.

Reply:
The [`EquatableInterface`](http://api.symfony.com/4.0/Symfony/Component/Security/Core/User/EquatableInterface.html) looks like to be determined to be implemented to catch up with stuff like role changes (because role changes are not detected by the [`AbstractToken::hasUserChanged`](https://github.com/symfony/symfony/blob/3.2/src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php#L244-L287). 

But I understand your POV stated here:
> But the main point is that you can tailor the identitys runtime lifecycle to play nice with symfony without hacking your user entities lifecycle (especially when you want to use doctrine for the user entity).

Reply:
I think we resolved the entity issue in the Mumble call. The main issue would be that a Contao model restored from the session would not be in the registry, potentially leading to duplicate entities. Same applies in Doctrine and the UnitOfWork. However, Contao users are **not** entities, so that issue actually does not apply.